main.tf

```diff
@@ -1,13 +1,25 @@
 locals {
   is_t_instance_type = replace(var.instance_type, "/^t(2|3|3a){1}\\..*$/", "1") == "1" ? true : false
+  hostname_prefix    = var.hostname_prefix != "" ? var.hostname_prefix : "ip-`curl -s 169.254.169.254/latest/meta-data/local-ipv4/ | sed -e 's/\\./-/g'`"
+  hostnames          = formatlist("%s-%d", local.hostname_prefix, range(1, var.instance_count + 1))
+}
+
+data "template_file" "user_data" {
+  count = var.instance_count
+
+  template = var.user_data
+
+  vars = {
+    hostname = local.hostnames[count.index]
+  }
 }
 
 resource "aws_instance" "this" {
   count = var.instance_count
 
   ami              = var.ami
   instance_type    = var.instance_type
-  user_data        = var.user_data
+  user_data        = data.template_file.user_data[count.index].rendered
```

Comment:
How about doing the following approach with an external file, because [the doc says they don't recommend an inline template string, and and it mentions `templatefile` function](https://www.terraform.io/docs/providers/template/d/file.html) ?

```
user_data = templatefile("${path.module}/user_data.tmpl", { hostname: local.hostnames[count.index] })
```

Reply:
Thanks!
I know this `templatefile` function built in from 0.12. But personally i don't want to add tmpl file.
Because the forked module using user_data parameter, I don't want to significantly change existing features as much as possible.
 

Reply:
Thank you for the comment. 😀I understand you don't want to create an external file, but with `templatefile` the only change needed is this line, except for an external template file. So I think there is lower risk of conflicts in the future. 🤔

If we don't go with an external template, still I don't think the `user_data` variable should be processed as a template string because the name doesn't describe it is a template. (At least we should change the variable name.)

Alternatively, as long as we need only a short `user_data` script, a simple [string template](https://www.terraform.io/docs/configuration/expressions.html#string-templates) like the following should work. (I haven't tested.)
```
user_data = <<EOF
#!/bin/sh
hostnamectl set-hostname ${local.hostnames[count.index]}
EOF
```

Reply:
At first I tried the following but it didn't worked for me. 😅 

```
user_data = <<EOF
#!/bin/sh
hostnamectl set-hostname ${local.hostnames[count.index]}
EOF
```

Reply:
@tei-k [The doc](https://www.terraform.io/docs/configuration/expressions.html#string-literals) says that:

> In both quoted and heredoc string expressions, Terraform supports template sequences that begin with ${ and %{.

I tried it and it worked. Can you check it again?

Reply:
あー、 `$${` これを試してもだめだった気がします〜（web consoleから見た感じだとちゃんと設定されてる）、
また以前言った覚えはありますが、user_dataはparameterから渡せる機能は変えたくないなと思いまして現状となっていますね。（既存の機能をできれば変えたくない）